Abstractify the enemies

AnonymousFool opened the pull request:
New Assets:
- Assets/Scripts/Bomb.cs
  - New class associated with bomb enemies (used to be Enemy.cs)
  - Do nothing until player is within 100 units, then approach player
- Assets/Scripts/EnemyBehaviour.cs
  - Creates a new enum ENEMY_STATE that represents how aware an enemy is of the player
  - Creates a new Exception InvalidEnemyStateException to be thrown when an enemy has an unexpected ENEMY_STATE
  - Creates an abstract EnemyBehaviour class that provides framework for all enemy actions
  - Creates two extensions of EnemyBehaviour, DoNothing and ApproachTarget, both of which are used by bomb enemies

Major Changes:
- Assets/Scripts/Enemy.cs
  - No longer associated just with bomb enemies. Now an abstract class providing framework for general enemy types

AnonymousFool committed:
Renamed Enemy to Bomb and added empty abstract class Enemy.cs extending Monobehaviour and extended by Bomb.cs

AnonymousFool committed:
Added MoveBehaviours and CombatBehaviours

Micky774 commented on Assets/Scripts/CombatBehaviour.cs:
```diff
@@ -0,0 +1,21 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class CombatBehaviour
+{
+    public MoveBehaviour moveBehaviour;
+    
+    public abstract void Act();
+}
```
For reference I think the two classes `CombatBehaviour` and `MoveBehaviour` should be disentagled fully. I don't `CombatBehaviour` should _rely_ on `MoveBehaviour`.

AnonymousFool replied:
lol I'm gonna need an enemy more complex than this Bomb before we can even test an implementation then. It literally doesn't have a CombatBehaviour distinct from its MoveBehaviour under your suggested framework, it just closes distance.

Micky774 replied:
> lol I'm gonna need an enemy more complex than this Bomb before we can even test an implementation then. It literally doesn't have a CombatBehaviour distinct from its MoveBehaviour under your suggested framework, it just closes distance.

I know, I expect that they'll both have the same functionality right now. It's mainly a matter of API design.

AnonymousFool replied:
I'm going to disentangle them by renaming CombatBehaviour to AttackBehaviour, because at a guess, that's more what you were thinking. That being said, I'm also going to remove "JustMove" as an AttackBehaviour because it doesn't really make sense, that's not an attack.

Micky774 replied:
So here's my thinking. We can essentially factor generic AI into three phases:

1. Idle: no information on active threats, be it player or mob
2. Aware: is aware of a threat, and is either avoiding or seeking
3. Engaged: is directly engaged with threat, and is either acting offensively/defensively

The most trivial rule is just "do nothing" for all three. A _slightly_ less trivial rule is to do nothing for `idle`, move directly towards enemy in `aware` and move directly towards enemy in `engage`. That's the behavior our sample enemies -- the bombs -- display right now. For debugging purposes, it's sufficient to use `Debug.Log` to confirm that we are in the correct phases during a gameplay test. I think we need to get comfortable using gameplay tests and debug logs to confirm our implementations.

Each of these, in my opinion, should be as independent as possible. There may be some which have very limited interplay (e.g. idle behavior A is _only_ used with aware behaviors B or C and engage behavior D...) but from a code-organization perspective it should still be factored separately because in the general case I think we expect reusability of these behaviors.

Alternatively, if we really do expect that each monster is going to be unique, we can simply have methods `Combat{Idle, Aware, Engage}' and reimplement them in each `Enemy` subclass, factoring out reusable code (e.g. direct movement as we currently have) into a combat-utility namespace (similarly to how we handle angle computation).

What do you think?

AnonymousFool replied:
Okay, the `idle`, `aware`, `engaged` trichotomy makes a lot more sense than just `MoveBehaviour` and `AttackBehaviour` dichotomy to me, especially since movement is frequently part of attacking, and `engaged` is going to look like fleeing for some enemies. I also don't think we should expect that each monster is going to be wholly unique, I'm very sure we're going to be able to make `idle` behaviours that can work for different enemies just by tweaking a couple of instance variables.

I do wonder though if it makes sense to have three different abstract classes for these kinds of behaviours though. Like we clearly have a case with the `Bomb` where an enemy has the same `aware` behaviour as their `engaged` behaviour, and it seems like it might just make more sense to make an abstract `EnemyBehaviour` class that all these behaviours extend and then give the abstract `Enemy` class 3 `EnemyBehaviour` instance variables, say `idleAction`, `awareAction`, and `engagedAction`. We could make an `enum ENEMY_STATE` that takes on values `idle`, `aware`, and `engaged`, then give the `Enemy` abstract class an instance variable storing its current state from the 3 `ENEMY_STATE` options and a basic `Update` implementation that updates its state variable (maybe using an abstract method `DetermineState` whose implementation varies by enemy) and then uses a simple switch to pick between the three actions based on that state.

The merit of this idea to me is that it enforces the trichotomy on the level where the distinction between the behaviours actually manifests, which is literally just how the enemy chooses which of its behaviours it engages in. We already have the Bomb whose `engaged` behaviour is the same as its `aware` behaviour, and we're just now discussing a mimic-like enemy in the art-dev channel who could very well have the same `DoNothing` behaviour for both the `idle` and `aware` states.

AnonymousFool committed:
Created new EnemyBehaviour framework

AnonymousFool commented:
I figure this goes without saying, but don't try to run this branch in Unity right now, scripts still have errors due to WIP code.

That being said, feel free to provide feedback on the current draft. I'm working with several things I'm not very familiar with rn, so there's probably quite a few improvements to be made that are more obvious to someone with a bit more experience.

AnonymousFool committed:
Added Bomb EnemyBehaviours

AnonymousFool committed:
Made a draft implementation of Enemy abstract class

AnonymousFool committed:
Updated scene for testing EnemyBehaviour

AnonymousFool commented:
Okay, I've updated the implementation such that everything compiles. Currently there's a debug message you can use to see the state change of the one enemy in the scene. iirc, I have it set up so that the enemy is engaged when the distance to the player is less than 50, aware between 50 and 100, and idle otherwise.

AnonymousFool committed:
Merge branch 'main' into enemy_abcs

AnonymousFool committed:
Fixed long-standing enemy update issue

Because enemies use Rigidbody2Ds, any changes to their velocity should be handled on FixedUpdate, which is tied to the physics tickrate. Despite correctly using Time.fixedDeltaTime, we've previously been updating enemies' velocity on Update instead of FixedUpdate. This commit fixes that.

AnonymousFool committed:
Added error checking, comments, and serialization

AnonymousFool committed:
Merge branch 'main' into enemy_abcs

Micky774 commented on Assets/Scripts/Bomb.cs:
```diff
@@ -4,22 +4,26 @@
         this._idleBehaviour = new DoNothing(this, ENEMY_STATE.idle);
-        this._awareBehaviour = new ApproachTarget(this, _player, _MAX_SPEED, _ACCELERATION_RATE, ENEMY_STATE.aware);
-        this._engagedBehaviour = new ApproachTarget(this, _player, _MAX_SPEED, _ACCELERATION_RATE, ENEMY_STATE.engaged);
+        this._awareBehaviour = new ApproachTarget(this, _player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.aware);
+        this._engagedBehaviour = new ApproachTarget(this, _player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.engaged);
```
I'm surprised `Bomb._{MAX_SPEED, ACCELERATION_RATE}` works granted they're not static variables. Shouldn't we use `this._XXX` instead?

AnonymousFool replied:
> I'm surprised `Bomb._{MAX_SPEED, ACCELERATION_RATE}` works granted they're not static variables. Shouldn't we use `this._XXX` instead?

they're const, which means they're tied to the class rather than the object.

AnonymousFool committed:
Fixed exception catch not outputting exception message

Micky774 commented on Assets/Scripts/Enemy.cs:
```diff
@@ -2,34 +2,49 @@
+    // We assume all enemies will have a Rigidbody2D for collision purposes
+    protected Rigidbody2D _myRigidBody;
+
+    // We enforce that an enemy define a Start method to initialize its enemy-specific behaviours
+    protected abstract void Start();
```
Can't we omit this since it is inherited from `MonoBehavior` anyways?

AnonymousFool committed:
Added error-checking to PerformAction
Also updated comments.

AnonymousFool committed:
Formatted document automatically with vscode for stylistic consistency

AnonymousFool commented:
Just updated initial comment to list the major changes.

I also haven't looked into making a custom editor for enemies yet, I forgor before hitting the re-review button.

Micky774 committed:
Updated LTS editor version

AnonymousFool committed:
Merge branch 'update_version' into enemy_abcs

AnonymousFool committed:
Merge branch 'main' into enemy_abcs

Micky774 commented:
> Just updated initial comment to list the major changes.
> 
> I also haven't looked into making a custom editor for enemies yet, I forgor before hitting the re-review button.
> 
> Update: I think serializing by reference is currently the better solution simply because it looks like if I made a custom editor that let us change behaviour attributes in the inspector for specific enemies, it would (a) require a lot more work and (b) probably have about the same effective computational hit just due to the low complexity of the current `EnemyBehaviour` classes.
> 
> I don't have much evidence of either of these claims (although `[SerializeReference]` is only one line, so it's probably a lot less work). For now, our CPU main stat says it's still between 2 and 3 ms. If we reach the point that we need to optimize, this is one of the first things I'll experiment with; we're just not at a point in the project yet where I can get actual data on which strategy is truly more computationally efficient.

Shall we just inline a `TODO` comment on researching potential alternatives?

AnonymousFool commented:
> Shall we just inline a `TODO` comment on researching potential alternatives?

Sounds like a great idea. I'll write one and push it in a bit.

AnonymousFool committed:
Added TODO comment on [SerializeReference]

Micky774 commented on Assets/Scripts/Enemy.cs:
```diff
@@ -2,34 +2,66 @@
+public abstract class Enemy : MonoBehaviour
 {
+    // Each enemy state has a corresponding behaviour the enemy performs while in that state
+    /* 
+     * TODO: Research if there's a more computationally efficient way to allow editing of EnemyBehaviour members in Inspector than serializing the EnemyBehaviours by reference.
+     * Example possibilities include defining a custom Editor subclass to add EnemyBehaviour members to Inspector
+     * As of 5/6/2023, the project is small enough that the overhead of [SerializeReference] is negligible. This is impermanent and experimentation will be required.
```
```suggestion
    /* 
     * TODO: Research if there's a more computationally efficient way to allow editing of EnemyBehaviour members in Inspector than serializing the EnemyBehaviours by reference.
     * Example possibilities include defining a custom Editor subclass to add EnemyBehaviour members to Inspector.
```

AnonymousFool replied:
Interesting. Do we not want to timestamp the most recent time someone looks into this problem?

Micky774 replied:
In general I don't see the need. If we decide to revisit this tomorrow or one year from now, it should be dictated by the state of the project rather than an arbitrary timeline. We don't even timestamp on really large projects like scikit-learn.

AnonymousFool replied:
I disagree with the conceptual justification, as I don't find the inclusion of a timestamp to add reliance on an arbitrary timeline, it just adds information that can be evaluated at any later point about the non-arbitrary timeline choice of... the timeline that has happened on the project thus far. It seems like you hold the belief that by timestamping it, it means the same thing in six months (i.e. oh my god we haven't looked at this in six months) regardless of the timeline that actually occurs, hence the emphasis on the arbitrariness of the timeline. But for those of us with moderately good memories of things from six months ago, it provides a reference point to look back to and compare the current state of the project against to evaluate whether the project is now **different** in any key ways that make the TODO more immediately actionable. In essence, it lets us consider not just the current position, but the deltas, which in my opinion does add potentially valuable information to the decision-making process.

That all being said, I legitimately have no preference in this specific case. I don't think the timestamp is going to be particularly necessary for me because my memory is slightly better than moderately good, and I'm often capable of remembering the state of the project in which I last considered an issue like this **without** the timestamp. You clearly have a stronger preference here toward what you consider normal project organization practice, so I'm going to commit the suggestion.

Micky774 commented on Assets/Scripts/Enemy.cs:
```diff
@@ -2,34 +2,66 @@
+    {
+        switch (this._state)
+        {
+            case ENEMY_STATE.idle:
+                return _idleBehaviour.Act();
+            case ENEMY_STATE.aware:
+                return _awareBehaviour.Act();
+            case ENEMY_STATE.engaged:
+                return _engagedBehaviour.Act();
```
```suggestion
            case ENEMY_STATE.idle:
                return this._idleBehaviour.Act();
            case ENEMY_STATE.aware:
                return this._awareBehaviour.Act();
            case ENEMY_STATE.engaged:
                return this._engagedBehaviour.Act();
```

AnonymousFool replied:
Good call, missed that.

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,197 @@
+    /// </summary>
+    idle = 0,
+    /// <summary>
+    /// Enemy is aware of player but not engaged in conflict
```
Just to try to keep the language consistent

```suggestion
    /// Enemy is aware of player but not engaged in combat
```

AnonymousFool committed:
Update Assets/Scripts/Enemy.cs

Co-authored-by: Meekail Zain <34613774+Micky774@users.noreply.github.com>

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,197 @@
+    /// Throws an exception with a given error message.
+    /// </summary>
+    /// <param name="message"> Error message </param>
+    public InvalidEnemyStateException(string message) : base(message)
+    {
+    }
```
For empty bodies, I prefer the following style:
```suggestion
    public InvalidEnemyStateException(string message) : base(message) {}
```

What do you think?

AnonymousFool replied:
Yeah actually, I prefer having it on one line like you just edited it.

Would you be okay if I put a space between the curly braces? The advantage of that is just that vscode automatically puts a space there if you use its "format document" feature, which I've been using lately just because it makes it easier to adhere to a more consistent curly brace style.

Micky774 replied:
Sure

AnonymousFool replied:
Awesome. In that case, I just pushed this as a change in a recent update, so I'll resolve this for now since I've indirectly accepted the suggestion.

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,197 @@
+    /// <exception cref="InvalidEnemyStateException"> Thrown if state is not an intended use case for the EnemyBehaviour </exception>
+    protected void _CheckUseCase(ENEMY_STATE? state)
+    {
+        if (state is ENEMY_STATE state_val
+        && !this._use_case[(int)state_val])
```
```suggestion
        if (
            state is ENEMY_STATE state_val
            && !this._use_case[(int)state_val]
         )
```

AnonymousFool replied:
Trying to use parentheses consistently like curly braces?

If so, did you mean to indent the closing parenthesis an extra space? it's in the column with the "f" of "if" instead of the "i".

Micky774 replied:
Did not mean the extra space

AnonymousFool committed:
Update Assets/Scripts/EnemyBehaviour.cs

Co-authored-by: Meekail Zain <34613774+Micky774@users.noreply.github.com>

AnonymousFool committed:
Update Assets/Scripts/Enemy.cs

Co-authored-by: Meekail Zain <34613774+Micky774@users.noreply.github.com>

AnonymousFool committed:
Fixed curly braces for InvalidEnemyStateError

AnonymousFool committed:
Changed spacing on multi-line if-statement

AnonymousFool committed:
Merge branch 'main' into enemy_abcs

AnonymousFool committed:
Updated omnisharp.json to reflect MUGA standard for curly braces

AnonymousFool committed:
Revert "Updated omnisharp.json to reflect MUGA standard for curly braces"

Pushed to the wrong branch due to absentmindedness

This reverts commit 194b241e7477dc0ced9279715ecd34d97e448867.

AnonymousFool committed:
Formatted new scripts to adhere to MUGA

Micky774 commented on Assets/Scripts/Bomb.cs:
```diff
@@ -0,0 +1,39 @@
+    // Bomb aware of player within 100 units, engaged within 50.
+    protected override void ChangeState() {
+        float dist_to_player = (this._myRigidbody.position - this._player.position).magnitude;
+        if(dist_to_player < 50) {
+            this._state = ENEMY_STATE.engaged;
+        } else if(dist_to_player < 100) {
+            this._state = ENEMY_STATE.aware;
+        } else {
+            this._state = ENEMY_STATE.idle;
+        }
```
This is functionally equivalent but I think it is helpful to put the emphasis on the `idle` state being the default
```suggestion
        ENEMY_STATE state = ENEMY_STATE.idle;
        if(dist_to_player < 50) {
            state = ENEMY_STATE.engaged;
        } else if(dist_to_player < 100) {
            state = ENEMY_STATE.aware;
        } 
        this._state = state;
```

Micky774 commented on Assets/Scripts/Bomb.cs:
```diff
@@ -0,0 +1,39 @@
+            this._state = ENEMY_STATE.idle;
+        }
+
+        // Use the following line of code to check that _state changes
+        // Debug.Log("My distance to the player is " + dist_to_player + " so my current enemy state is " + this._state);
```
I wonder if we could instead use some global debug pre-processor macro, e.g.
```
#if DEBUG
Debug.Log("DEBUG");
#endif
```

that way we can enable/disable it on the project-level instead of individually via comments

AnonymousFool replied:
damn it, deleted my response here by mistake.

what if instead we used a public boolean instance variable that defaults to false, but can be changed in the Inspector? That way, if we know we want something in debug mode at runtime beforehand, we can change it before running, and also if we realize mid-runtime that we need some debug output logged due to unexpected behaviour, then we can change it at runtime.

If you want, I could also make a script tied to the Enemies gameobject (the parent of all the individual enemy objects) that has its own debug instance variable that can be checked. Then if that parent debug variable is true when an enemy performs its start method, I'll have the enemy switch its own debug variable to true so that all enemies in the project will be in debug mode with one variable change.

Micky774 replied:
I want to avoid using runtime checks since they are expensive. That branching can be predicted easily usually, but I imagine w/ how Unity calls on many different hooks each tick/frame, the branch likely won't reside on cache long enough to be optimized by the CPU, making it costly. Meanwhile macros are determined at compile-time, so during run-time they incur no cost.

AnonymousFool replied:
Is the runtime cost not worth the benefit of the ability to reactively enable debug mode mid-runtime? If we isolate debug code to `Start()` and `FixedUpdate()`, then we're doing a maximum of one `if`-statement per entity on their first frame and a maximum of 1 additional `if`-statement per entity per physics tick.

I'd say that the benefit likely outweighs the cost. One `if`-statement is likely not going to be the bottleneck of a `FixedUpdate` for an `Enemy`, and being able to enable debug output at runtime is a super-useful feature. If you still think the cost is significant enough to be worth considering, than I can run testing like we did with switches and dictionaries to see just how much more computationally efficient the macros are, but I don't think that one `if`-statement per `FixedUpdate` is going to be the straw that breaks the CPU's back.

AnonymousFool replied:
We discussed this in discord and decided to go for pre-processor macros due to the scalability issues of lots of conditionals being normalized across the project for debugging purposes due to problems like branch-misses causing suboptimal performance and memory management.

AnonymousFool replied:
Alright, I added a pre-processor macro solution in place of the stupid comment solution I was using before. Suggestion implemented.

Micky774 commented on Assets/Scripts/Bomb.cs:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    // Bomb aware of player within 100 units, engaged within 50.
+    protected override void ChangeState() {
```
I think we should instead name this something like `GetState` or `UpdateState` so as not imply that the outgoing state is guaranteed to be different from the original state

Micky774 commented on Assets/Scripts/Enemy.cs:
```diff
@@ -2,34 +2,57 @@
+        if(this._can_change_state) {
+            this.ChangeState();
+        }
+        try {
+            this._can_change_state = this.PerformAction();
+        } catch(InvalidEnemyStateException ex) {
+            Debug.LogException(ex);
+            this._state = ENEMY_STATE.idle;
+            this._can_change_state = true;
+        }
```
Is this really what we want to happen? Catching/handling exceptions implies we tolerate it and it is within the realm of expected behavior. Isn't this error indicative of something which is a relatively critical mistake in the backend that should be rectified directly, rather than tolerated and compensated for? Formally, shouldn't we allow this error to propagate and interrupt rather than catch and handle it directly

Micky774 commented on Assets/Scripts/Enemy.cs:
```diff
@@ -2,34 +2,57 @@
-        myRigidbody.velocity = Vector2.ClampMagnitude(myRigidbody.velocity + acceleration, max_speed);
+    // PerformAction is virtual to allow overriding by custom non-state-based enemies (such as bosses, potentially)
+    protected virtual bool PerformAction() {
+        switch(this._state) {
+            case ENEMY_STATE.idle:
+                return this._idleBehaviour.Act();
+            case ENEMY_STATE.aware:
+                return this._awareBehaviour.Act();
+            case ENEMY_STATE.engaged:
+                return this._engagedBehaviour.Act();
+            default:
+                throw new InvalidEnemyStateException("INVALID ENEMY STATE ERROR: " + this.GetType() + " has a non-standard enemy state while using base PerformAction function.");
+        }
```
imo we should prefer using `Dictonary` types over switches and complex conditionals whenever possible. Any way to avoid branching is preferred for readability and performance.

Sometimes branching is necessary, when the logical operations required are genuinely different and cannot be factored out, but here I think we can safely and effectively replace branching.

AnonymousFool replied:
Replaced switch with dictionary lookup. Suggestion accepted.

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,180 @@
+    /// <param name="enemy"> Enemy who will use the DoNothing behaviour </param>
+    /// <param name="state"> ENEMY_STATE in which the DoNothing behaviour will be used. Null to skip use-case checking </param>
+    public DoNothing(Enemy enemy, ENEMY_STATE? state) {
+        this._use_case[(int)ENEMY_STATE.idle] = true;
+        this._use_case[(int)ENEMY_STATE.aware] = true;
```
May we abstract into a small helper with the following signature: `this._set_use_cases(case1, case2, ...)` or alternatively `this._set_use_cases([case1, case2, ...])`

AnonymousFool replied:
Absolutely. Done.

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,180 @@
+
+        // Calculates change in velocity for current frame
+        Vector2 delta_v = this._target.position - this._enemyBody.position;
+        this._enemySprite.flipX = delta_v.x <= 0;
```
Since different enemies way want different cosmetics even for the same action, I'm inclined to think we should care for those in the actual appropriate `Enemy` subclass, maybe with a similar `After<behavior>Animate` hook

AnonymousFool replied:
Is there any reason to use events to handle this instead of just making abstract AfterIdleAnimate, AfterAwareAnimate, and AfterEngagedAnimate methods that we implement in the subclass?

I thought at first that you wanted to use events for Animator compatibility, but the Animator uses a separate trigger system with boolean parameters that you set from inside C# scripts, so that's definitely not the motivating reason.

AnonymousFool replied:
I just learned what a hook is. I'm great at this.

AnonymousFool replied:
Alright this is now implemented, so this specific issue is resolved, but I'm going to raise a follow-up issue because of a choice I made in the implementation that I want your eyes on.

AnonymousFool committed:
Added XML comments to Bomb.cs and Enemy.cs

AnonymousFool committed:
Update Assets/Scripts/Bomb.cs

Co-authored-by: Meekail Zain <34613774+Micky774@users.noreply.github.com>

AnonymousFool committed:
Renamed ChangeState to UpdateState

AnonymousFool committed:
Removed error-catching on fatal InvalidEnemyStatException

AnonymousFool committed:
Replaced PerformAction switch with dictionary lookup

AnonymousFool committed:
Added _SetUseCase helper method

Micky774 commented on Assets/Scripts/EnemyBehaviour.cs:
```diff
@@ -0,0 +1,180 @@
+        if (
+            state is ENEMY_STATE state_val
+            && !this._use_case[(int)state_val]
+        ) {
```
Yeah this more aligns w/ the styles we've talked about. The lack of it originally was mainly due to how it started and that I didn't make my suggestion capture the line w/ the curly brace :P

Either way, looks great like this :D

AnonymousFool committed:
Added preprocessor directive for enemy debug code

AnonymousFool committed:
Refactored animations to occur in new functions

These animation functions are implemented on an enemy-by-enemy basis

AnonymousFool commented on Assets/Scripts/Enemy.cs:
```diff
@@ -1,35 +1,114 @@
+    /// <summary>
+    /// Dictionary of EnemyBehaviours and animation functions to be used by Enemy in associated ENEMY_STATE
+    /// </summary>
+    protected IDictionary<ENEMY_STATE, (EnemyBehaviour, Action)> _behaviours = new Dictionary<ENEMY_STATE, (EnemyBehaviour, Action)>();
```
I've now had to change the dictionary to one that stores not just EnemyBehaviours, but tuples of EnemyBehaviours and Actions. This is because now we need to take an ENEMY_STATE and perform **not just** a corresponding EnemyBehaviour **but also** an animation function.

I had two obvious choices in how I could implement this. I could select the animation function using a switch on ENEMY_STATE values or I could use a dictionary lookup. I defaulted to a dictionary lookup for readability, and in order to perform only one dictionary lookup instead of two, I expanded the values of our preexisting dictionary instead of making a second new dictionary.

AnonymousFool replied:
I forgot to @Micky774 even though you're the one whose eyeballs I need on this comment and the corresponding comment below.

AnonymousFool replied:
This is no longer the current implementation, and as such is outdated and resolved.

AnonymousFool commented on Assets/Scripts/Bomb.cs:
```diff
@@ -0,0 +1,87 @@
+        this._player = GameObject.FindGameObjectWithTag("Player").GetComponent<Rigidbody2D>();
+        this._behaviours.Add(ENEMY_STATE.idle, (new DoNothing(this, ENEMY_STATE.idle), AfterIdleAnimate));
+        this._behaviours.Add(ENEMY_STATE.aware, (new ApproachTarget(this, this._player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.aware), AfterAwareAnimate));
+        this._behaviours.Add(ENEMY_STATE.engaged, (new ApproachTarget(this, this._player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.engaged), AfterEngagedAnimate));
```
The consequence of how I changed the dictionary structure is now adding elements to the dictionary is kinda complicated. I'd make a helper function, but the only thing it could do without itself using a switch to decide which animation function to use based on the ENEMY_STATE is to make it so you can pass the ENEMY_STATE value, the EnemyBehaviour, and the animation function as three separate parameters instead of an ENEMY_STATE value and a tuple of the other two parameters. Which means you call a helper function so you don't have to write two parentheses. Is that even worth it?

Micky774 replied:
We could create a new class that serves as a data structure for this pattern

AnonymousFool replied:
Are you suggesting we make a wrapper class `EnemyData` that stores an `EnemyBehaviour` and an animation function (probably in the form of an `Action`)? And then make `behaviours` a `Dictionary<ENEMY_STATE, EnemyData>`, and in individual enemy classes like `Bomb` we split up the process of adding an element to `behaviours` across two lines, say

`EnemyData idleData = new EnemyData(new DoNothing(this, ENEMY_STATE.idle), AfterIdleAnimate);
this._behaviours.Add(ENEMY_STATE.idle, idleData);`

for greater code readability?

Micky774 replied:
Aye

AnonymousFool replied:
Since we had this conversation, we refactored Bomb.cs Start method to make the use of the structure significantly more readable, and we've made minor tweaks to the struct's API to make much more convenient to use.

As I have gotten the approval I so deeply craved, I'm resolving this.

AnonymousFool committed:
Added EnemyData struct for readability

AnonymousFool committed:
Added XML comments to EnemyData structure

AnonymousFool committed:
Changed two needlessly public variables to protected

Also added one newline character to Bomb.cs

AnonymousFool committed:
Comment fixes

AnonymousFool committed:
Changed bool[] _use_case to ENEMY_STATE[] _use_cases

AnonymousFool committed:
Changed exception-handling for _CheckUseCase

Now instead of a try-catch block, _CheckUseCase will output the exception message directly to the console if an InvalidEnemyStateException occurs.
This should make EnemyBehaviour subclass writing easier.

AnonymousFool committed:
Fixed XML comment for _CheckUseCase

AnonymousFool committed:
Renamed acceleration_dir to unscaled_acc

This rename clarifies that the vector is not a unit vector.

AnonymousFool committed:
Changed EnemyBehaviour (and subclass) constructor parameters

Now instead of passing the Enemy as a parameter to EnemyBehaviour subclass constructors, we pass just the Enemy's Rigidbody2D. This means we don't have to call GetComponent<Rigidbody2D> in those constructors, saving computation time.
Further, I've moved the assignment of the _enemyBody instance variable to a new base EnemyBehaviour constructor to make the code cleaner.

AnonymousFool committed:
Refactored delta_v calculation for improved readability

AnonymousFool committed:
Exposed EnemyBehaviour Act method through EnemyData API

AnonymousFool committed:
Refactored Bomb Start() method code for readability

AnonymousFool committed:
Changed API expectations for _behaviours in Enemy

We now clearly expect that developers ensure the _behaviours has an EnemyData corresponding to each ENEMY_STATE value _state achieves in an Enemy subclass.
This allows us to simply conditional checks that would otherwise need to occur on every physics tick.

AnonymousFool committed:
Fixed dangling modifier in Enemy Act() summary comment

AnonymousFool committed:
Updated XML comments for clarity

AnonymousFool committed:
Made aware and engaged distances into instance variables

This allows easier tweaking at runtime to find good distances.

AnonymousFool committed:
Merge branch 'main' into enemy_abcs

AnonymousFool committed:
More merge conflict resolution

AnonymousFool committed:
Removed enemy debug flag.
